File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.cs
using System;
using System.Diagnostics.CodeAnalysis;
using Alias = System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverageAttribute;

[ExcludeFromCodeCoverage] // Noncompliant ^2#23 {{Add a justification.}}
class Noncompliant
{
    [ExcludeFromCodeCoverage()] // Noncompliant
    void WithBrackets() { }

    [System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage] // Noncompliant
    void FullyDeclaredNamespace() { }

    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage] // Noncompliant
    void GloballyDeclaredNamespace() { }

    [ExcludeFromCodeCoverage(Justification = null)] // Noncompliant
    void WithNull() { }

    [ExcludeFromCodeCoverage(Justification = "")] // Noncompliant
    void WithEmptyString() { }

    [ExcludeFromCodeCoverage(Justification = "  ")] // Noncompliant
    void WithWhiteSpace() { }

    [Alias(Justification = "")] // Noncompliant
    void WithAlias() { }

    [ExcludeFromCodeCoverage] // Noncompliant
    [CLSCompliant(false)]
    uint Multiple() { return 0; }

    [ExcludeFromCodeCoverage, CLSCompliant(false)]
//   ^^^^^^^^^^^^^^^^^^^^^^^
    uint Combined() { return 0; }

    [ExcludeFromCodeCoverage] // Noncompliant
    Noncompliant() { }

    [ExcludeFromCodeCoverage] // Noncompliant
    void Method() { }

    [ExcludeFromCodeCoverage] // Noncompliant
    event EventHandler Event;
}

interface IInterface
{
    [ExcludeFromCodeCoverage] // Noncompliant
    void Method();
}

[ExcludeFromCodeCoverage] // Noncompliant
struct ProgramStruct
{
    [ExcludeFromCodeCoverage] // Noncompliant
    void Method() { }
}

[ExcludeFromCodeCoverage(Justification = "justification")]
class Compliant
{
    [ExcludeFromCodeCoverage(Justification = "justification")]
    Compliant() { }

    [ExcludeFromCodeCoverage(Justification = "justification")]
    void Method() { }

    [ExcludeFromCodeCoverage(Justification = "justification")]
    string Property { get; set; }

    [ExcludeFromCodeCoverage(Justification = "justification")]
    event EventHandler Event;
}

interface IComplaintInterface
{
    [ExcludeFromCodeCoverage(Justification = "justification")]
    void Method();
}

[ExcludeFromCodeCoverage(Justification = "justification")]
struct ComplaintStruct
{
    [ExcludeFromCodeCoverage(Justification = "justification")]
    void Method() { }
}

class NotApplicable
{
    [CLSCompliant(false)]
    enum Enum { foo, bar }

    NotApplicable() { }

    void Method() { }

    int Property { get; set; }

    int Field;

    event EventHandler Event;

    delegate void Delegate();

    [OtherNamespace.ExcludeFromCodeCoverage]
    void SameName() { }
}

namespace OtherNamespace
{
    public class ExcludeFromCodeCoverageAttribute : Attribute { }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExcludeFromCodeCoverageAttributesNeedJustification.vb
' Commented line for concurrent namespace
Imports System.Diagnostics.CodeAnalysis
Imports [Alias] = System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverageAttribute

<ExcludeFromCodeCoverage> ' Noncompliant^2#23 {{Add a justification.}}
Class Noncompliant
    <ExcludeFromCodeCoverage()> Sub WithBrackets() ' Noncompliant^6#25
    End Sub

    <System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage> ' Noncompliant
    Sub FullyDeclaredNamespace()
    End Sub

    <Global.System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage> ' Noncompliant
    Sub GloballyDeclaredNamespace()
    End Sub

    <[Alias](Justification:="")> ' Noncompliant
    Sub WithAlias()
    End Sub

    <ExcludeFromCodeCoverage(Justification:=Nothing)> ' Noncompliant
    Sub WithNothing()
    End Sub

    <ExcludeFromCodeCoverage(Justification:="")> ' Noncompliant
    Sub WithEmptyString()
    End Sub

    <ExcludeFromCodeCoverage(Justification:="  ")> ' Noncompliant
    Sub WithWhiteSpace()
    End Sub

    <ExcludeFromCodeCoverage> ' Noncompliant
    <CLSCompliant(False)>
    Function Multiple() As UInteger
        Return 0
    End Function

    <ExcludeFromCodeCoverage, CLSCompliant(False)> Function Combined() As UInteger ' Noncompliant
        Return 0
    End Function

    <ExcludeFromCodeCoverage> ' Noncompliant
    Sub New()
    End Sub

    <ExcludeFromCodeCoverage> ' Noncompliant
    Sub Method()
    End Sub

    <ExcludeFromCodeCoverage> ' Noncompliant
    Property [Property] As Integer

    <ExcludeFromCodeCoverage> ' Noncompliant
    Event [Event] As EventHandler

End Class

Interface IInterface
    <ExcludeFromCodeCoverage> ' Noncompliant
    Sub Method()
End Interface

<ExcludeFromCodeCoverage> ' Noncompliant
Structure ProgramStruct
    <ExcludeFromCodeCoverage> ' Noncompliant
    Sub Method()
    End Sub
End Structure

<ExcludeFromCodeCoverage(Justification:="justification")>
Class Compliant

    <ExcludeFromCodeCoverage(Justification:="justification")>
    Sub New()
    End Sub

    <ExcludeFromCodeCoverage(Justification:="justification")>
    Sub Method()
    End Sub

    <ExcludeFromCodeCoverage(Justification:="justification")>
    Property [Property] As String

    <ExcludeFromCodeCoverage(Justification:="justification")>
    Event [Event] As EventHandler
End Class

Interface IComplaintInterface
    <ExcludeFromCodeCoverage(Justification:="justification")>
    Sub Method()
End Interface

<ExcludeFromCodeCoverage(Justification:="justification")>
Structure ComplaintStruct
    <ExcludeFromCodeCoverage(Justification:="justification")>
    Sub Method()
    End Sub
End Structure

Class NotApplicable

    Sub New()
    End Sub

    Sub Method()
    End Sub

    Property [Property] As Integer

    Event [Event] As EventHandler

    <NotSystem.ExcludeFromCodeCoverageAttribute>
    Sub SameName()
    End Sub
End Class

Namespace NotSystem
    Public Class ExcludeFromCodeCoverageAttribute
        Inherits Attribute
    End Class
End Namespace
